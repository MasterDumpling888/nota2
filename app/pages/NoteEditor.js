import React, { useState, useEffect } from 'react';
import { View, Text, TextInput, Button, StyleSheet, TouchableOpacity, FlatList, ScrollView } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import NoteNavBar from '../components/NoteNavBar';
import PageBox from '../components/PageBox';
import Icon from 'react-native-vector-icons/Ionicons';
import markdownStyles from '../markdownStyles';
import Markdown from 'react-native-markdown-display';
import { getFontSize } from '../responsiveFont';
import { addNote, getNote, updateNote, fetchFolders, fetchTags } from '../services/notesService';
import { summarizeNote } from '../services/aiService';
import Footer from '../components/Footer';

function NoteEditor({ route, navigation }) {
  const { noteId } = route.params || {};
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');
  const [summarizedContent, setSummarizedContent] = useState('');
  const [folder, setFolder] = useState('');
  const [tag, setTag] = useState('');
  const [folders, setFolders] = useState([]);
  const [tags, setTags] = useState([]);
  const [folderDropdownVisible, setFolderDropdownVisible] = useState(false);
  const [tagDropdownVisible, setTagDropdownVisible] = useState(false);

  useEffect(() => {
    const loadNote = async () => {
      if (noteId) {
        try {
          const note = await getNote(noteId);
          setTitle(note.title);
          setContent(note.content);
          setSummarizedContent(note.summary || '');
          setFolder(note.folder || '');
          setTag(note.tag || '');
        } catch (error) {
          console.error('Failed to load note', error);
        }
      }
    };

    const loadFolders = async () => {
      try {
        const foldersList = await fetchFolders();
        setFolders(foldersList);
      } catch (error) {
        console.error('Failed to load folders', error);
      }
    };

    const loadTags = async () => {
      try {
        const tagsList = await fetchTags();
        setTags(tagsList);
      } catch (error) {
        console.error('Failed to load tags', error);
      }
    };

    loadNote();
    loadFolders();
    loadTags();
  }, [noteId]);

  const handleSaveNote = async () => {
    try {
      const note = {
        title,
        content,
        folder,
        tag,
        summary: summarizedContent
      };

      if (noteId) {
        await updateNote(noteId, note);
      } else {
        await addNote(note);
      }
      navigation.goBack();
    } catch (error) {
      console.error('Failed to save note', error);
    }
  };

  const handleSummarizeNote = async () => {
    try {
      const summary = await summarizeNote(content);
      setSummarizedContent(summary);
    } catch (error) {
      console.error('Failed to summarize note', error);
    }
  };

  const renderFolderItem = ({ item }) => (
    <TouchableOpacity onPress={() => { setFolder(item.id); setFolderDropdownVisible(false); }}>
      <Text style={styles.dropdownItem}>{item.name}</Text>
    </TouchableOpacity>
  );

  const renderTagItem = ({ item }) => (
    <TouchableOpacity onPress={() => { setTag(item.id); setTagDropdownVisible(false); }}>
      <Text style={styles.dropdownItem}>{item.name}</Text>
    </TouchableOpacity>
  );

  return (
    <SafeAreaView style={{ flex: 1 }}>
      <PageBox title={noteId ? "Edit Note" : "Create Note"} onClose={() => navigation.goBack()}>
        <ScrollView showsVerticalScrollIndicator={false}>
          <View style={styles.header}>
            <TextInput
              style={styles.input}
              value={title}
              placeholder='Add your title'
              onChangeText={setTitle}
            />
            <TouchableOpacity style={styles.aiButton} onPress={handleSummarizeNote} >
              <Icon name="sparkles" size={24} color="black" style={styles.closeIcon} />
            </TouchableOpacity>
          </View>
          <View style={styles.previewContainer}>
            <Text style={styles.inputTitle}>Summary:</Text>
            <Text style={styles.outputContent}>{summarizedContent}</Text>
          </View>
          <View style={styles.previewContainer}>
            <Text style={styles.inputTitle}>Preview:</Text>
            <Text style={styles.outputContent}>
              <Markdown style={markdownStyles}>{content}</Markdown>
            </Text>
          </View>
          <TextInput
            style={styles.textArea}
            value={content}
            placeholder='Start typing here! (Markdown supported)'
            onChangeText={setContent}
            multiline
          />

          <View style={styles.inputContainer}>
            <Text style={styles.inputTitle}>Folder</Text>
            <TouchableOpacity onPress={() => setFolderDropdownVisible(!folderDropdownVisible)} style={styles.selector}>
              <Text style={styles.selectorText}>{folder ? folders.find(f => f.id === folder)?.name : 'Select Folder'}</Text>
              <Icon name="caret-down" size={24} color="#33FD0A" />
            </TouchableOpacity>
            {folderDropdownVisible && (
              <View style={styles.dropdown}>
                <FlatList
                  data={folders}
                  renderItem={renderFolderItem}
                  keyExtractor={(item) => item.id}
                />
              </View>
            )}
          </View>
          <View style={styles.inputContainer}>
            <Text style={styles.inputTitle}>Tag</Text>
            <TouchableOpacity onPress={() => setTagDropdownVisible(!tagDropdownVisible)} style={styles.selector}>
              <Text style={styles.selectorText}>{tag ? tags.find(c => c.id === tag)?.name : 'Select Tag'}</Text>
              <Icon name="caret-down" size={24} color="#33FD0A" />
            </TouchableOpacity>
            {tagDropdownVisible && (
              <View style={styles.dropdown}>
                <FlatList
                  data={tags}
                  renderItem={renderTagItem}
                  keyExtractor={(item) => item.id}
                />
              </View>
            )}
          </View>
          <TouchableOpacity style={styles.saveButton} onPress={handleSaveNote} >
            <Text style={{ fontFamily: 'Raleway-SemiBold', fontSize: getFontSize(18) }}>Save</Text>
          </TouchableOpacity>
          <Footer />
        </ScrollView>
      </PageBox>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  input: {
    fontSize: getFontSize(24),
    fontFamily: 'Raleway-SemiBold',
    color: 'white',
  },
  inputTitle: {
    fontSize: getFontSize(18),
    fontFamily: 'Raleway-SemiBold',
    color: 'white',
    marginBottom: 8,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingRight: 32,
  },
  textArea: {
    backgroundColor: '#33FD0Aa3',
    borderRadius: 5,
    padding: 8,
    marginBottom: 16,
    height: 'auto',
    color: 'white',
    fontSize: getFontSize(16),
    fontFamily: 'Raleway-Regular',
  },
  outputContent: {
    color: 'white',
    fontSize: getFontSize(16),
    fontFamily: 'Raleway-Regular',
  },
  saveButton: {
    padding: 16,
    backgroundColor: '#33FD0A',
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 16,
  },
  aiButton: {
    padding: 8,
    backgroundColor: '#33FD0A',
    borderRadius: 100,
    justifyContent: 'center',
    alignItems: 'center',
  },
  selector: {
    borderWidth: 2,
    borderColor: '#33FD0A',
    borderRadius: 5,
    padding: 8,
    marginVertical: 8,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  selectorText: {
    color: 'white',
    fontSize: getFontSize(16),
    fontFamily: 'Raleway-Regular',
  },
  dropdown: {
    borderWidth: 1,
    borderColor: '#33FD0A',
    borderRadius: 5,
    marginVertical: 8,
    maxHeight: 150,
  },
  dropdownItem: {
    padding: 8,
    borderBottomWidth: 1,
    backgroundColor: '#33FD0Aa3',
    borderBottomColor: '#33FD0A',
    color: 'white',
    fontFamily: 'Raleway-SemiBold',
    fontSize: getFontSize(14),
  },
  previewContainer: {
    marginTop: 16,
    marginBottom: 16,
    borderLeftWidth: 2,
    borderLeftColor: '#33FD0A',
    padding: 8,
  },
});

export default NoteEditor;
