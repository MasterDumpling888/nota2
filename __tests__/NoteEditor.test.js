import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import NoteEditor from '../app/pages/NoteEditor';

// Mock services specific to NoteEditor
jest.mock('../app/services/notesService', () => ({
    addNote: jest.fn(),
    getNote: jest.fn(() => Promise.resolve({
        title: 'Test Note',
        content: 'Test Content',
        folder: 'folder1',
        tag: 'tag1'
    })),
    updateNote: jest.fn(),
    fetchFolders: jest.fn(() => Promise.resolve([
        { id: 'folder1', name: 'Personal' },
        { id: 'folder2', name: 'Work' }
    ])),
    fetchTags: jest.fn(() => Promise.resolve([
        { id: 'tag1', name: 'Important' },
        { id: 'tag2', name: 'Urgent' }
    ]))
}));

jest.mock('../app/services/aiService', () => ({
    summarizeNote: jest.fn(() => Promise.resolve('This is a summarized note'))
}));

// Mock SafeAreaView
jest.mock('react-native-safe-area-context', () => ({
    SafeAreaView: 'SafeAreaView'
}));

// Mock components not already in jest.setup.js
jest.mock('../app/components/Footer', () => 'Footer');
jest.mock('../app/components/NoteNavBar', () => 'NoteNavBar');

describe('NoteEditor Component', () => {
    const mockNavigation = {
        navigate: jest.fn(),
        goBack: jest.fn()
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('renders correctly for editing an existing note', async () => {
        const mockRoute = {
            params: {
                noteId: '123'
            }
        };

        const { getByText } = render(
            <NoteEditor route={mockRoute} navigation={mockNavigation} />
        );
    });

    test('renders correctly for creating a new note', async () => {
        const emptyRoute = { params: {} };

        const { getByPlaceholderText } = render(
            <NoteEditor route={emptyRoute} navigation={mockNavigation} />
        );

        await waitFor(() => {
            // Check for placeholder text for a new note
            expect(getByPlaceholderText('Add your title')).toBeTruthy();
            expect(getByPlaceholderText('Start typing here! (Markdown supported)')).toBeTruthy();
        });
    });

    test('NoteEditor snapshot test', () => {
        const mockRoute = {
            params: {
                noteId: '123'
            }
        };

        const { toJSON } = render(
            <NoteEditor route={mockRoute} navigation={mockNavigation} />
        );
        expect(toJSON()).toMatchSnapshot();
    });
});