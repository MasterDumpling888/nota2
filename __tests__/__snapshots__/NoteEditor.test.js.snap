// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditor Component NoteEditor snapshot test 1`] = `
<SafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <PageBox
    onClose={[Function]}
    title="Edit Note"
  >
    <RCTScrollView
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingRight": 32,
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Add your title"
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 24,
              }
            }
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#33FD0A",
                "borderRadius": 100,
                "justifyContent": "center",
                "opacity": 1,
                "padding": 8,
              }
            }
          >
            <Icon
              color="black"
              name="sparkles"
              size={24}
            />
          </View>
        </View>
        <View
          style={
            {
              "borderLeftColor": "#33FD0A",
              "borderLeftWidth": 2,
              "marginBottom": 16,
              "marginTop": 16,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 18,
                "marginBottom": 8,
              }
            }
          >
            Summary:
          </Text>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-Regular",
                "fontSize": 16,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderLeftColor": "#33FD0A",
              "borderLeftWidth": 2,
              "marginBottom": 16,
              "marginTop": 16,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 18,
                "marginBottom": 8,
              }
            }
          >
            Preview:
          </Text>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-Regular",
                "fontSize": 16,
              }
            }
          >
            <MarkdownDisplay
              style={
                {
                  "blockquote": {
                    "backgroundColor": "#646464",
                    "borderLeftColor": "#33FD0A",
                    "borderLeftWidth": 4,
                    "color": "white",
                    "fontFamily": "Raleway-Italic",
                    "fontSize": 16,
                    "marginVertical": 10,
                    "paddingLeft": 10,
                    "textAlign": "left",
                  },
                  "body": {
                    "color": "white",
                    "fontSize": 16,
                    "textAlign": "auto",
                  },
                  "bullet_list_icon": {
                    "marginLeft": 3,
                    "marginRight": 3,
                  },
                  "code_inline": {
                    "backgroundColor": "#646464",
                    "color": "#33FD0A",
                    "fontFamily": "Courier",
                    "fontSize": 16,
                  },
                  "em": {
                    "color": "white",
                    "fontFamily": "Raleway-Italic",
                    "textAlign": "left",
                  },
                  "fence": {
                    "backgroundColor": "#646464",
                    "borderColor": "#33FD0A",
                    "borderLeftColor": "#33FD0A",
                    "borderLeftWidth": 4,
                    "borderRadius": 3,
                    "color": "#33FD0A",
                    "fontFamily": "Courier",
                    "fontSize": 16,
                    "marginVertical": 10,
                    "padding": 10,
                  },
                  "heading1": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 24,
                  },
                  "heading2": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 20,
                  },
                  "heading3": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 18,
                  },
                  "heading4": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 16,
                  },
                  "heading5": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 14,
                  },
                  "heading6": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "fontSize": 12,
                  },
                  "link": {
                    "color": "#33FD0A",
                  },
                  "list_item": {
                    "color": "white",
                    "flexDirection": "row",
                    "fontFamily": "Raleway-Regular",
                    "fontSize": 14,
                    "justifyContent": "flex-start",
                    "width": "100%",
                  },
                  "ordered_list_icon": {
                    "marginLeft": 3,
                    "marginRight": 3,
                  },
                  "paragraph": {
                    "color": "white",
                    "fontFamily": "Raleway-Regular",
                    "fontSize": 16,
                  },
                  "strong": {
                    "color": "white",
                    "fontFamily": "Raleway-Bold",
                    "textAlign": "left",
                  },
                  "table": {
                    "borderColor": "#33FD0A",
                    "borderRadius": 3,
                    "borderWidth": 1,
                    "marginVertical": 10,
                  },
                  "tbody": {
                    "fontFamily": "Raleway-Regular",
                    "fontSize": 12,
                  },
                  "td": {
                    "borderColor": "#33FD0A",
                    "borderWidth": 1,
                    "color": "white",
                    "flex": 1,
                    "fontFamily": "Raleway-Regular",
                    "padding": 5,
                  },
                  "th": {
                    "borderColor": "#33FD0A",
                    "borderWidth": 1,
                    "color": "white",
                    "flex": 1,
                    "fontFamily": "Raleway-Bold",
                    "padding": 5,
                  },
                  "thead": {
                    "backgroundColor": "#33FD0A",
                    "color": "black",
                  },
                  "tr": {
                    "borderBottomWidth": 1,
                    "borderColor": "#33FD0A",
                    "flexDirection": "row",
                  },
                }
              }
            />
          </Text>
        </View>
        <TextInput
          multiline={true}
          onChangeText={[Function]}
          placeholder="Start typing here! (Markdown supported)"
          style={
            {
              "backgroundColor": "#33FD0Aa3",
              "borderRadius": 5,
              "color": "white",
              "fontFamily": "Raleway-Regular",
              "fontSize": 16,
              "height": "auto",
              "marginBottom": 16,
              "padding": 8,
            }
          }
          value=""
        />
        <View>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 18,
                "marginBottom": 8,
              }
            }
          >
            Folder
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "borderColor": "#33FD0A",
                "borderRadius": 5,
                "borderWidth": 2,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 8,
                "opacity": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontFamily": "Raleway-Regular",
                  "fontSize": 16,
                }
              }
            >
              Select Folder
            </Text>
            <Icon
              color="#33FD0A"
              name="caret-down"
              size={24}
            />
          </View>
        </View>
        <View>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 18,
                "marginBottom": 8,
              }
            }
          >
            Tag
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "borderColor": "#33FD0A",
                "borderRadius": 5,
                "borderWidth": 2,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 8,
                "opacity": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontFamily": "Raleway-Regular",
                  "fontSize": 16,
                }
              }
            >
              Select Tag
            </Text>
            <Icon
              color="#33FD0A"
              name="caret-down"
              size={24}
            />
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#33FD0A",
              "borderRadius": 5,
              "justifyContent": "center",
              "marginTop": 16,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              {
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 18,
              }
            }
          >
            Save
          </Text>
        </View>
        <Footer />
      </View>
    </RCTScrollView>
  </PageBox>
</SafeAreaView>
`;
